src/internal/clusterstate/v2.8.0/clusterstate.go

```diff
@@ -32,5 +32,6 @@ func Migrate(state migrations.State) migrations.State {
 				return errors.Wrap(err, "could not synthesize cluster defaults")
 			}
 			return nil
-		})
+		}).
+		Apply("Synthesize user and effective specs from their pipeline details", synthesizeSpecs)
```

Comment:
Squash this with the previous `Synthesize cluster defaults from environment variables`
```suggestion
		Apply("Synthesize user and effective specs from their pipeline details", synthesizeSpecs, migration.Squash)
```